host/lib/usrp/dboard/e3xx/e3xx_radio_control_init.cpp

```diff
@@ -159,6 +164,58 @@ void e3xx_radio_control_impl::_init_frontend_subtree(
         .add_coerced_subscriber([](const std::vector<std::string>&) {
             throw uhd::runtime_error("Attempting to update antenna options!");
         });
+    // TX Power Calibration
+    auto tx_cal_ggroup = uhd::gain_group::make();
+    const std::string tx_ant = subtree->access<std::string>(tx_fe_path / "antenna/value").get();
+    tx_cal_ggroup->register_fcns(HW_GAIN_STAGE, {
+            [this, chan_idx]() { return get_tx_gain_range(chan_idx); },
+            [this, chan_idx]() { return get_tx_gain(chan_idx); },
+            [this, chan_idx](const double gain) { this->set_tx_gain(gain, chan_idx); }
+        },
+        10 /* High priority */);
+    _tx_pwr_mgr.insert(_tx_pwr_mgr.begin() + chan_idx,
+        uhd::usrp::pwr_cal_mgr::make(
+            // Cal serial is the DB serial plus the FE name
+            db_serial + "#" + subtree->access<std::string>(tx_fe_path / "name").get(),
+            // Logger prefix
+            "E3XX-CAL-TX",
+            // Current frequency getter
+            [this, chan_idx]() { return get_tx_frequency(chan_idx); },
+            // Current calibration key getter (format is "e3xx_pwr_$dir_$chan_$ant")
+            [chan_str = std::to_string(chan_idx),
+             ant_str = pwr_cal_mgr::sanitize_antenna_name(tx_ant)]() {
```

Comment:
I think you don't need the fourth commit at all, if you dynamically read the current antenna value in this lambda. See here: https://github.com/EttusResearch/uhd/blob/master/host/lib/usrp/b200/b200_impl.cpp#L1044-L1051

The B200 also uses different cal values for TX/RX and RX2 ports, and also only needs a single power cal manager object per channel.

Reply:
Here's how X300 does this: https://github.com/EttusResearch/uhd/blob/master/host/lib/usrp/x300/x300_radio_control.cpp#L1757-L1766

Reply:
And X410: https://github.com/EttusResearch/uhd/blob/master/host/lib/usrp/dboard/zbx/zbx_dboard_init.cpp#L270-L283

Reply:
Thanks for the feedback!  It may be another week or two until I can circle back and revisit this, but I haven't forgotten or lost interest in this PR..

Reply:
This looks good.